[PSR-15] Provide middleware dispatch examples

weierophinney opened the pull request:
Per discussions on #987, on the mailing list, and in the PSR-15 working group, this patch builds on #987 to add a new section 6.3 to the meta document, detailing "Example Interface Interactions".

The section covers the following:

- A queue-based request handler that aggregates and dispatches middleware.
- A decorator-based request handler for building a layered dispatch mechanism.
- Two examples of re-usable middleware to demonstrate de-coupling from the application and the request handler.
  
This patch incorporates feedback made on #987, but not incorporated there.

weierophinney commented:
Ping @Crell — I think the new section answers your questions, but would like for you to review and confirm.

Also pinging @shadowhand — do the changes look okay for you? Also, I rebased off of master, which picked up changes merged since you opened #987.

Crell commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+
+- Middleware does not need to know anything about any other middleware or how it
+  is composed in the application.
+- The `QueueRequestHandler` is agnostic of the PSR-7 implementation in use
```
Missing period on this line.

Crell commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+This system has essentially two request handlers, one that will produce a
+response if no middleware generates one, and one for dispatching the middleware
+layers. (In this example, the `RoutingMiddleware` will likely execute composed
+handlers on a successful route match; see more on that below.)
```
I don't really see 2 handlers here... I see one handler ($app) that will end up invoked 2-3 times.  Am I missing something or is this worded awkwardly?

weierophinney replied:
No, you're correct. The original example I created composed another request handler instead of a request prototype, and delegated to that handler if the last middleware in the queue calls on the handler instead of returning a response. I'm going to refactor the example to to do that, as it also then answers the "request prototype" question you have elsewhere.

Crell commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+    }
+}
+
+$innerHandler = new class ($responsePrototype) implements RequestHandlerInterface {
```
Nitpick: $responsePrototype isn't defined yet.  Is this supposed to be a `new Response())->withStatus(500)`?

weierophinney replied:
It could be, or a 404 status. I'll update to make this more clear.

Crell commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+Note the use of a "response prototype" in the above example; this approach
+allows the middleware to be agnostic of which PSR-7 implementation is in use,
+and yet still return a response in cases where it determines it should not
+delegate to the handler.Similarly, it is not concerned with how the request
```
Missing space on this line.

Crell commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+these to work in either situation, we need to write them such that they interact
+appropriately.
+
+Generally speaking, implementors of middleware should follow these guidelines:
```
I feel like these guidelines should be moved to the spec proper, in some form, as they directly inform how middlewares will interoperate.  Eg, when to call the next handler cannot be captured in an interface but is still relevant, and if done "wrong" will lead to a non-interoperable implementation.

weierophinney replied:
@shadowhand What do you think about moving these into the spec proper? If we do, where should they live?

shadowhand replied:
I don't think we should but putting code samples in the spec itself. Having the guidelines at the top of the spec would be good though.

Crell replied:
Yes, I was referring to the guidelines, (probably written a bit differently).  I agree the code samples should remain in the metadoc.

mnapoli replied:
I really think the spec should not dictate *how* middlewares should behave (or other implementation details), e.g. when to generate a response or call the delegate.

Maybe that's not what you meant, but the guidelines below are too precise IMO.

stefanotorresi replied:
I agree with @shadowhand and @mnapoli: I don't think the spec document should contain these guidelines.

weierophinney replied:
@mnapoli and @stefanotorresi — thanks; I think that's great feedback, and I tend to agree. The guidelines presented are implementation details and good practices, but _do not need to be followed in order to comply with the specification_. I'll leave them where they are.

mnapoli replied:
Could you make that part clear in the text? Because it reads like a specification (with the "should" keyword).

weierophinney replied:
I'm not sure what other word to user here, TBH, and the fact that it's not ALL CAPS means it does not have the same semantic weight per RFC 2119. Maybe the following?

> Implementors of middleware striving for maximum interoperability may want to consider the following guidelines:

Would that work for you, @mnapoli ?

weierophinney replied:
Updated!

Crell commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+
+        return $corsResult->isCorsScoped()
+            ? $corsResult->prepareResponse($this->responsePrototype)
+            : $response;
```
It's cool to see someone besides me that uses/likes that particular formatting idiom. :smile:

Crell commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -275,6 +276,10 @@ and a request handler and must return a response. The middleware may:
 
+It is expected that a middleware dispatching system will use the request handler
+to delegate response creation to the next available middleware. The last middleware
+would then act as a domain gateway to execute application code.
```
This paragraph... doesn't quite make sense.  I can't wrap my head around what it's saying.  Let me try to rephrase and tell me if I'm in the right ballpark:

```
When delegating from one middleware to the next in a sequence, a dispatching system should use an intermediary request handler as a way to link two middlewares together.  The final or innermost middleware acts as a gateway to application code that will, in most cases, be responsible for response generation.
```

Is that accurate?  (Feel free to steal liberally if so.)  Should that be in the main spec in some form?

weierophinney replied:
It's correct, except that the "application code that will, in most cases, be responsible for response generation" will likely either be the middleware itself, or the a _request handler_. I'll see if I can come up with a re-phrase that makes sense.

Crell commented:
This is a huge improvement, but still some bits aren't as direct as they could be.  See commentary above.

mnapoli commented on proposed/http-handlers/request-handlers.md:
```diff
@@ -110,20 +110,20 @@ use Psr\Http\Message\ResponseInterface;
+ *
+ * An HTTP middleware component participates in processing an HTTP message;
+ * for example, by acting on the request, generating the response, or forwarding the
+ * request to a subsequent middleware and possibly acting on its response.
```
👍

shadowhand commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+
+The two interfaces, `RequestHandlerInterface` and `MiddlewareInterface`, were
+designed to work in conjunction with one another. While a
+`RequestHandlerInterface` can feasibly be used in isolation, it gains more power
```
This sentence isn't necessary, imo. RequestHandler is perfectly viable *without* middleware, not the other way around.

shadowhand commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+  a new response by manipulating the one returned (e.g., `return
+  $response->withHeader('X-Foo-Bar', 'baz')`).
+
+The `CorsMiddleware` is one that will exercise all three of these guidelines: if
```
I am mildly uncomfortable using a highly specific example like this. This assumes that I know what CORS is and what the failure conditions mean. If we're going to use a specific example perhaps something like routing would be more generally understood?

weierophinney replied:
I'll see if I can come up with another example. This was one I liked because it hits all the various ways middleware might produce a response:

- on its own, due to a failed pre-condition
- by acting as a pass-through to the request handler
- by modifying the response returned by the request handler

Let me see if I can come up with an example, and I'll ping you if/when I do.

weierophinney replied:
Came up with an example using authorization instead of CORS; please review and let me know what you think.

shadowhand commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+handler is implemented; it merely uses it to produce a response when
+pre-conditions have been met.
+
+The `RoutingMiddleware` implementation described below follows a similar
```
Yes this is better, but I think the process should be:

- If routing matches, continue processing.
- Else return a `HTTP Not Found` response.

weierophinney replied:
That's actually exactly how we do it in Expressive. :+1:

The only issue I see with that is that it complicates the example: I then need to add dispatch middleware and a response prototype or factory. I felt the "map route to handler" approach was sufficiently simple and didn't require extra explanation.

shadowhand commented on proposed/http-handlers/request-handlers.md:
```diff
@@ -112,8 +112,8 @@ use Psr\Http\Message\ServerRequestInterface;
- * An HTTP middleware component participates in processing an HTTP message,
- * either by acting on the request, generating the response, or forwarding the
+ * An HTTP middleware component participates in processing an HTTP message;
+ * for example, by acting on the request, generating the response, or forwarding the
```
Why the change of phrasing? "Either" seems like the appropriate wording.

weierophinney replied:
This was based on feedback from @mnapoli, IIRC.

Crell replied:
This seems fine.  If we want to get super pedantic then "either" is technically only correct when there's 2 options, and in this case there are 3.  "For example" seems OK to me.

Crell replied:
Or, technically, with the semicolon we could omit that word entirely and it would still be a grammatically valid list of the available options.

weierophinney commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -276,9 +276,11 @@ and a request handler and must return a response. The middleware may:
+system should use an intermediary request handler as a way to link middleware
+together. The final or innermost middleware will act as a gateway to application
+code and generate a response from its results; alternately, the middleware may
+delegate this responsibility to a _request handler_.
```
@Crell This is the updated text, using your suggestion as a starting point.

Crell replied:
Much clearer!  Why is the last "request handler" set off, though?  Perhaps "to a dedicated request handler"?  (Nitpicking here.)

A statement of this sort seems like it belongs in the spec.  It's not a "why and background", it's a "no, really, this is how you're supposed to use it".  That seems like it should be in the spec proper.

weierophinney commented on proposed/http-handlers/request-handlers-meta.md:
````diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
+  the request to the handler, and sign the response returned based on the request.
+
+```php
+class AuthorizationMiddleware implements MiddlewareInterface
````
@shadowhand Let me know if this altered example (replacing CorsMiddleware with AuthorizationMiddleware) makes more sense. I definitely want to demonstrate multiple middleware, and particularly different approaches middleware can take both to be re-usable as well as to change the workflow/paths of an application.

shadowhand commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
+    {
+        if (! $authorizationMap->requestRequiresAuthorization($request)) {
```
How about `doesRequireAuthorization` instead of `requestRequiresAuthorization`?

Jean85 replied:
Maybe `needsAuthorization` to keep it short? (for easiness of readability)

shadowhand commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
+        $this->responsePrototype = $responsePrototype;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
```
`): ResponseInterface` as per purposed PSR-12 spec, no space.

https://github.com/php-fig/fig-standards/blob/master/proposed/extended-coding-style-guide.md#45-method-and-function-arguments

shadowhand commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
+        ResponseInterface $responsePrototype
+    ) {
+        $this->authorizationMap = $authorizationMap;
+        $this->responsePrototype = $responsePrototype;
```
I'm not sure why we need a response prototype here. I think it would make more sense of the `AuthorizationMap` was responsible for generating/manipulating the response. That would simplify the example and keep the focus on the middleware.

weierophinney replied:
Agreed; making that change now.

Crell commented on proposed/http-handlers/request-handlers-meta.md:
```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
+    public function __construct(
+        AuthorizationMap $authorizationMap,
+        ResponseInterface $responsePrototype
+    ) {
```
There's only 2 parameters. No need to go with the harder-to-read multi-line approach.  Icky.

shadowhand replied:
When using objects as type, I prefer the multi-line approach. Easier to read and less churn on diffs when parameters change.

weierophinney replied:
I was going to disagree, but when I re-formatted, I realized the line is still shorter than the `process()` method declaration line. About to commit that now.

weierophinney replied:
Plus, with the change to remove the response prototype, it's now only one argument anyways...

weierophinney commented:
Requested changes have been made (mostly):

- I **did not** move the "guidelines" bullet points to the spec; as several folks noted, coding to those guidelines will definitely mean you're coding to the spec, but you do not need to follow them to implement the spec.
- I removed the emphasis around "request handler", and added the word "dedicated" before them, per @Crell.
- I moved that particular paragraph to section 1.2 of the spec, per @Crell.
- I updated the preamble to `MiddlewareInterface` to incorporate the various bits of feedback.
- I updated code samples to follow PSR-12.
- I made changes to the `AuthorizationMiddleware` example per feedback received.

@Crell, @shadowhand, @mnapoli and @stefanotorresi — can you give one more review so we can finalize these? I think we can then prepare for a CC vote.

Crell commented on proposed/http-handlers/request-handlers.md:
```diff
@@ -52,6 +52,12 @@ the creation of a resulting response, as defined by PSR-7.
+system should use an intermediary request handler as a way to link middleware
+together. The final or innermost middleware will act as a gateway to application
+code and generate a response from its results; alternately, the middleware may
+delegate this responsibility to a dedicated request handler.
```
Nitpick: SHOULD and MAY should be capitalized to mean the defined SHOULD and MAY, I think, since this is in the spec proper.

weierophinney replied:
My understanding is they should only be capitalized when they have the semantic meanings as defined in RFC 2119; the question is whether or not those words here have those semantic meanings. I did not originally write this with those meanings in mind.

The problem, of course, is that there are very few ways to re-word this to have the same intended meaning _without_ using those words. Can you propose an alternative? Or do you feel they have those semantics?

ashnazg replied:
If not RFC2119, maybe `can` could be used for both?

weierophinney replied:
`CAN` also has semantic meaning via RFC 2119. :)

Been discussing this and the idea of moving the section to the spec document itself heavily today; hopefully we can have resolution shortly.

ashnazg replied:
Hmm, I don't see `CAN` mentioned in the RFC... hence me thinking it safe.

Would something wordy, like `could potentially`, do the job without sounding like a 2119 directive?

Crell commented on proposed/http-handlers/request-handlers-meta.md:
````diff
@@ -368,6 +364,233 @@ executing that handler. If unable to route to a handler, it would instead
+from the outside-in, passing each request handler "layer" to the next outer one.
+
+```php
+class DecoratedRequestHandler implements RequestHandlerInterface
````
Nitpick: Technically, this isn't a decoratED handler.  It's a decoratING handler.  The name should be more descriptive.  Sorry for not catching that before.

weierophinney replied:
Fixed.

weierophinney commented:
At this time, based on discussion with the working group, I've moved the paragraph in question back to the meta document's section 6.2 preamble, with changes as discussed here and elsewhere to remove implications that these are anything more than optional recommendations.
